sympy/physics/quantum/state.py

```diff
@@ -617,6 +619,59 @@ def dual_class(self):
         return TimeDepKet
 
 
+class OrthogonalState(State, StateBase):
+    """General abstract quantum state used as a base class for Ket and Bra."""
+    pass
+
+class OrthogonalKet(OrthogonalState, KetBase):
+    """Orthogonal Ket in quantum mechanics.
+
+    The inner product of two states with different labels will give zero,
+    states with the same label will give one.
+
+        >>> from sympy.physics.quantum import OrthogonalBra, OrthogonalKet
+        >>> OrthogonalBra(n)*OrthogonalKet(n)
+        1
+        >>> OrthogonalBra(n)*OrthogonalKet(n+1)
+        0
+        >>> OrthogonalBra(n)*OrthogonalKet(m)
+        <n|m>
+    """
+
+    @classmethod
+    def dual_class(self):
+        return OrthogonalBra
+
+    def _eval_innerproduct(self, bra, **hints):
+
+        if len(self.args) != len(bra.args):
+            raise ValueError('Cannot multiply a ket that has a different number of labels.')
+
+        for i in range(len(self.args)):
+            diff = self.args[i] - bra.args[i]
+            diff.simplify()
```

Comment:
Maybe this should really be `diff = diff.simplify()`. It's probably better not to simplify here. The user can simplify afterwards.

Reply:
The simplification is there just to increase the chances of recognising a constant-zero expressions. The result of simplification is not returned to the user and since `diff` is only an internal variable, the user can't easily simplify it themselves.
But I'm not sure about how good are `is_zero` and `is_nonzero` at managing complicated expressions. Are they guaranteed not to change during simplification?

Reply:
What kind of object is diff? Does its simplify method work in place?

Reply:
`diff` is an expresssion (specifically an `Add` node), you're right about `diff = diff.simplify()` being needed. I fixed that in the latest commit.